core/src/main/java/org/pentaho/di/core/vfs/KettleVFS.java

```diff
@@ -0,0 +1,608 @@
+    if ( scheme == null ) {
+      return fileSystemOptions;
+    }
+
+    return buildFsOptions( space, fileSystemOptions, vfsFilename, scheme );
+  }
+
+  /**
+   * Private method for stripping password from filename when a FileObject
+   * can not be obtained.
+   * getFriendlyURI(FileObject) or getFriendlyURI(String) are the public
+   * methods.
+   */
+  private static String cleanseFilename( String vfsFilename ) {
+    return vfsFilename.replaceAll( ":[^:@/]+@", ":<password>@" );
+  }
+
+  private static FileSystemOptions buildFsOptions( VariableSpace varSpace, FileSystemOptions sourceOptions,
+                                                   String vfsFilename, String scheme ) throws IOException {
+    if ( varSpace == null || vfsFilename == null ) {
+      // We cannot extract settings from a non-existant variable space
+      return null;
+    }
+
+    IKettleFileSystemConfigBuilder configBuilder =
+      KettleFileSystemConfigBuilderFactory.getConfigBuilder( varSpace, scheme );
+
+    FileSystemOptions fsOptions = ( sourceOptions == null ) ? new FileSystemOptions() : sourceOptions;
+
+    String[] varList = varSpace.listVariables();
+
+    for ( String var : varList ) {
+      if ( var.equalsIgnoreCase( CONNECTION ) && varSpace.getVariable( var ) != null ) {
+        FileSystemOptions fileSystemOptions = VFSHelper.getOpts( vfsFilename, varSpace.getVariable( var ) );
+        if ( fileSystemOptions != null ) {
+          return fileSystemOptions;
+        }
+      }
+      if ( var.startsWith( "vfs." ) ) {
+        String param = configBuilder.parseParameterName( var, scheme );
+        String varScheme = KettleGenericFileSystemConfigBuilder.extractScheme( var );
+        if ( param != null ) {
+          if ( varScheme == null || varScheme.equals( "sftp" ) || varScheme.equals( scheme ) ) {
+            configBuilder.setParameter( fsOptions, param, varSpace.getVariable( var ), var, vfsFilename );
+          }
+        } else {
+          throw new IOException( "FileSystemConfigBuilder could not parse parameter: " + var );
+        }
+      }
+    }
+    return fsOptions;
+  }
+
+  /**
+   * Read a text file (like an XML document). WARNING DO NOT USE FOR DATA FILES.
+   *
+   * @param vfsFilename the filename or URL to read from
+   * @param charSetName the character set of the string (UTF-8, ISO8859-1, etc)
+   * @return The content of the file as a String
+   * @throws IOException
+   */
+  public static String getTextFileContent( String vfsFilename, String charSetName ) throws KettleFileException {
+    return getTextFileContent( vfsFilename, null, charSetName );
+  }
+
+  public static String getTextFileContent( String vfsFilename, VariableSpace space, String charSetName )
+    throws KettleFileException {
+    try {
+      InputStream inputStream = null;
+
+      if ( space == null ) {
+        inputStream = getInputStream( vfsFilename );
+      } else {
+        inputStream = getInputStream( vfsFilename, space );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *inputStream*. The resource is referenced at lines 296, 299. The resource is closed at lines 302, 303. There are other execution paths that do not contain closure statements, for example, when InputStreamReader.read throws an exception. Ensure *inputStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).